command/agent/agent.go

```diff
@@ -600,6 +608,67 @@ func (a *Agent) setupClient() error {
 	return nil
 }
 
+// setupNodeID will pull the persisted node ID, if any, or create a random one
+// and persist it.
+func (a *Agent) setupNodeID(config *Config) error {
+	// If they've configured a node ID manually then just use that, as
+	// long as it's valid.
+	if config.NodeID != "" {
+		if _, err := uuid.ParseUUID(string(config.NodeID)); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// For dev mode we have no filesystem access so just make a GUID.
+	if a.config.DevMode {
+		id, err := uuid.GenerateUUID()
+		if err != nil {
+			return err
+		}
+
+		config.NodeID = types.NodeID(id)
+		a.logger.Printf("[INFO] agent: Generated unique node ID %q for this agent (will not be persisted in dev mode)", config.NodeID)
+		return nil
+	}
+
+	// Load saved state, if any. Since a user could edit this, we also
+	// validate it.
+	fileID := filepath.Join(config.DataDir, "node-id")
+	if _, err := os.Stat(fileID); err == nil {
+		rawID, err := ioutil.ReadFile(fileID)
+		if err != nil {
+			return err
+		}
+
+		nodeID := strings.TrimSpace(string(rawID))
+		if _, err := uuid.ParseUUID(nodeID); err != nil {
+			return err
+		}
+
+		config.NodeID = types.NodeID(nodeID)
+	}
+
+	// If we still don't have a valid node ID, make one.
+	if config.NodeID == "" {
+		id, err := uuid.GenerateUUID()
```

Comment:
This would be the place where we could hook into `github.com/shirou/gopsutil`:

```
import "github.com/shirou/gopsutil"
info, err := host.Info()
if info != nil && info.HostID != "" {
  config.NodeID = types.NodeID(info.HostID)
}
```

If the ID came from `gopsutil` then I wouldn't necessarily want to persist it in the data directory, but there shouldn't be any harm in that either since the ID coming from `gopsutil` and the OS should outlive the contents of whatever is in the data directory.